src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1571,6 +1571,180 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+                                                Register midx, Register rtmp,
+                                                int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    Label case0, case1, case2, case3;
+    Label *larr[] = {&case0, &case1, &case2, &case3};
+    for (int i = 0; i < 4; i++) {
+      // dst[i] = mask ? src[offset + index[i]] : 0
+      btq(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    Label case0, case1, case2, case3, case4, case5, case6, case7;
+    Label *larr[] = {&case0, &case1, &case2, &case3,
+                     &case4, &case5, &case6, &case7};
+    for (int i = 0; i < 8; i++) {
+      // dst[i] = mask ? src[offset + index[i]] : 0
+      btq(mask, midx);
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  }
+}
+#endif // _LP64
+
+void C2_MacroAssembler::vgather8b(BasicType elem_bt, XMMRegister dst,
+                                  Register base, Register idx_base,
+                                  Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      // dst[i] = src[index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      // dst[i] = src[index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather8b_offset(BasicType elem_bt, XMMRegister dst,
+                                         Register base, Register idx_base,
+                                         Register offset, Register rtmp,
+                                         int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      // dst[i] = src[offset + index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      // dst[i] = src[offset + index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+/*
+ * Gather loop first packs 4 short / 8 byte values from gather indices
+ * into quadword lane and then permutes quadword lane into appropriate
+ * location in destination vector. Following pseudo code describes the
```

Comment:
I found this a bit confusing. Would this be another way of saying this:
`We divide the gather operations into quadwords of 8 bytes (or 4 shorts). Each quadword is gathered using the vgather8b instruction. We combine the quadwords into the destination register via OR and shifting (i.e. rotating / permuting) the destination register, until every quadword is in the desired location.`